### PR TITLE
Java sdk

### PR DESCRIPTION
**Pull request by philippart-s**
@scraly need to talk with you if it's the right way to update and add Java SDK to the provider 😉

### CODE REVIEW

**Comment by philippart-s** [comment]
@scraly th files to check:
 - Makefile
 - build.gradle
 - settings.gradle

All other files are unmodified after auto generation.

Next steps : 
 - add the code and do a first integration?
 - need examples?
 - need docs?
 - ...?

**Comment by scraly** [comment]
Thanks for working on it 💪.

Preambule : if we want Java 21 we need to edit gitpod and devpod environments to jump java 11 to 21 :).


1. Makefile

To have a fully PR we need to have a Makefile that will generate the Java SDK without any manual modifications.

`build.gradle`and `settings.gradle` files can't be modified manually (because of the upgrade feature for instance) so a tips can be to edit them "automatically" in the Makefile in the `build_java` target.

Here an edited version of the `build_java` target (that replace Java version 11 to 21 and meta information in the `build/gradle` file and the rootProject in `settings.gradle` file):

```
build_java:: PACKAGE_VERSION := $(shell pulumictl get version --language generic)
build_java:: $(WORKING_DIR)/bin/$(JAVA_GEN)
	$(WORKING_DIR)/bin/$(JAVA_GEN) generate --schema provider/cmd/$(PROVIDER)/schema.json --out sdk/java  --build gradle-nexus
	rm -f ./provider/cmd/$(PROVIDER)/schema-java.json
	#echo "update java version in build.gradle" && cd ./sdk/java/ && sed -i -e 's/of(11)/of(21)/g' build.gradle
	echo "update meta info in build.gradle" && cd ./sdk/java/ && sed -i -e 's/info.metaClass.name = .*/info.metaClass.name = "$(JAVA_ARTIFACT_ID)"/g' \
                -e 's/group = .*/group = "$(JAVA_GROUP_ID)"/g' \
                -e 's/groupId = .*/groupId = "$(JAVA_GROUP_ID)"/g' \
                -e 's/artifactId = .*/artifactId = "$(JAVA_ARTIFACT_ID)"/g' \
                -e 's/description = .*/description = "A Pulumi package for creating and managing OVH resources."/g' ./build.gradle
	echo "update rootProject in settings.gradle" && cd ./sdk/java && sed -i -e 's/rootProject.name = .*/rootProject.name = "$(JAVA_GROUP_ID)"/g' settings.gradle
	cd sdk/java/ && \
		echo "module fake_java_module // Exclude this directory from Go tools\n\ngo 1.17" > go.mod && \
		gradle --console=plain build
```

And for the homogenous of the file, I suggest to add after `install_go_sdk` target:

```
install_java_sdk::
```

(empty like in Go)

And add it in `install_sdk` target:

```
install_sdks:: install_dotnet_sdk install_python_sdk install_nodejs_sdk install_java_sdk
```

2. .pulumi-java-gen.version

I see that the version is `0.11.0`, you can update it to `0.13.0` (latest version)

3. release.yaml

After having generated the Java SDK, we can publish (through a GitHub Action) the package.

To do that we need to:
- have an account in Maven central (https://central.sonatype.org/register/central-portal/)
- edit the `.github.workflows.release.yaml` file:

After `PYPI_USERNAME`:

```
  SIGNING_KEY: ${{ secrets.JAVA_SIGNING_KEY }}
  SIGNING_KEY_ID: ${{ secrets.JAVA_SIGNING_KEY_ID }}
  SIGNING_PASSWORD: ${{ secrets.JAVA_SIGNING_PASSWORD }}
  PUBLISH_REPO_PASSWORD: ${{ secrets.OSSRH_TOKEN_PASSWORD }}
  PUBLISH_REPO_USERNAME: ${{ secrets.OSSRH_TOKEN }}
```

After `PUBLISH_NUGET`:

```
  PUBLISH_NRM: true
```

After `Setup Python`:

```
      - name: Setup Java
        uses: actions/setup-java@v3
        with:
          cache: gradle
          distribution: temurin
          java-version: ${{ matrix.javaversion }}
```

After `publish nuget package` step:

```
      - if: ${{ matrix.language == 'java' && env.PUBLISH_NRM == 'true' }}
        name: Set PACKAGE_VERSION to Env for Java SDK
        run: echo "PACKAGE_VERSION=$(pulumictl get version --language generic)" >>
          $GITHUB_ENV
      - if: ${{ matrix.language == 'java' && env.PUBLISH_NRM == 'true' }}
        name: Publish Java SDK
        uses: gradle/gradle-build-action@9bca466e2722baf9e77b4f929bbe0af18848791f
        with:
          arguments: publishToSonatype closeAndReleaseSonatypeStagingRepository
          build-root-directory: ./sdk/java
          gradle-version: 7.4.1
```

Add `java` in matrix language list:

```
        language:
          - nodejs
          - python
          - go
          - dotnet
          - java
```

At the end of the file:

```
        javaversion:
          - "21"
```

4. GitHub secrets

We are using new GH secrets in the GH action so we need to add them.

5. README file

After publishing through the GH action the Java SDK, update the `README` file:
- the "badge" linking to Maven central package
- Java installing section

like this:

```
Java

PLEASE NOTE: Java is currently in preview. For more details check out [pulumi java documentation](https://www.pulumi.com/docs/intro/languages/java/#how-to-use-java-with-pulumi).

To use from Java, add the dependency below to your `pom.xml` file:

<dependency>
    <groupId>com.ovh.pulumi</groupId>
    <artifactId>ovh</artifactId>
    <version>[0.1.0,)</version>
</dependency>
```

File: Makefile [hunk 1]
```diff
@@ -10,7 +10,7 @@ PROVIDER_PATH    := provider
 VERSION_PATH     := ${PROVIDER_PATH}/pkg/version.Version
 
 JAVA_GEN 		 := pulumi-java-gen
-JAVA_GEN_VERSION := v0.8.0
+JAVA_GEN_VERSION := v0.12.0
 TFGEN           := pulumi-tfgen-${PACK}
 PROVIDER        := pulumi-resource-${PACK}
 VERSION         := $(shell pulumictl get version)
```
Comment (scraly) [inline Makefile:16]:
```suggestion
VERSION         := $(shell pulumictl get version)
JAVA_GROUP_ID    := com.${ORG}.pulumi
JAVA_ARTIFACT_ID := ${ORG}
```

**Review by scraly — CHANGES_REQUESTED** [review]

**Comment by scraly** [comment]
When we will succesfully publish the Java package through the GH action, then we can add examples.
To do that I created an issue: #185  

We'll can also create an example in Java in [kubernetes](https://github.com/ovh/pulumi-ovh/tree/main/examples/kubernetes) folder.

File: Makefile [hunk 2]
```diff
@@ -10,7 +10,7 @@ PROVIDER_PATH    := provider
 VERSION_PATH     := ${PROVIDER_PATH}/pkg/version.Version
 
 JAVA_GEN 		 := pulumi-java-gen
-JAVA_GEN_VERSION := v0.8.0
+JAVA_GEN_VERSION := v0.12.0
```
Comment (scraly) [inline Makefile]:
```suggestion
JAVA_GEN_VERSION := v0.13.0
```

**Comment by philippart-s** [comment]
waouh, thanks for this exhaustive review 😊.
I'll check all your remarks and update the PR, just for the Java version: I'll check in the Pulumi's java generator if are using Java 11 or 21. I think it's better to follow their version.

Again thanks a lot.

**Review by scraly — APPROVED** [review]

**Pull request merged**